### ISSUE TITLE
CancellationException thrown + unhandled after Single subscription disposed (RxJava 2)

### ISSUE THREAD

**Issue by kassim** [trello/RxLifecycle#217]
I'm not quite sure how to handle this issue.

I have a subscription to a Single that is bound to the lifecycle of a fragment, which I may at some point have to cancel/dispose. Once this has happened, if the fragment is closed the exception is thrown to the system and the app crashes. See below for the stack trace

```
java.util.concurrent.CancellationException
     at io.reactivex.internal.operators.single.SingleTakeUntil$TakeUntilOtherSubscriber.onNext(SingleTakeUntil.java:150)
     at io.reactivex.internal.subscriptions.DeferredScalarSubscription.complete(DeferredScalarSubscription.java:133)
     at io.reactivex.internal.operators.single.SingleToFlowable$SingleToFlowableObserver.onSuccess(SingleToFlowable.java:63)
     at io.reactivex.internal.operators.observable.ObservableElementAtSingle$ElementAtObserver.onNext(ObservableElementAtSingle.java:91)
     at io.reactivex.internal.operators.observable.ObservableFilter$FilterObserver.onNext(ObservableFilter.java:52)
     at io.reactivex.internal.operators.observable.ObservableOnErrorReturn$OnErrorReturnObserver.onNext(ObservableOnErrorReturn.java:66)
     at io.reactivex.internal.operators.observable.ObservableCombineLatest$LatestCoordinator.drain(ObservableCombineLatest.java:246)
     at io.reactivex.internal.operators.observable.ObservableCombineLatest$LatestCoordinator.combine(ObservableCombineLatest.java:199)
     at io.reactivex.internal.operators.observable.ObservableCombineLatest$CombinerObserver.onNext(ObservableCombineLatest.java:316)
     at io.reactivex.internal.operators.observable.ObservableSkip$SkipObserver.onNext(ObservableSkip.java:53)
     at io.reactivex.internal.operators.observable.ObservableRefCount$ConnectionObserver.onNext(ObservableRefCount.java:169)
     at io.reactivex.internal.operators.observable.ObservablePublish$PublishObserver.onNext(ObservablePublish.java:233)
     at io.reactivex.internal.util.NotificationLite.accept(NotificationLite.java:246)
     at io.reactivex.subjects.BehaviorSubject$BehaviorDisposable.test(BehaviorSubject.java:490)
     at io.reactivex.subjects.BehaviorSubject$BehaviorDisposable.emitNext(BehaviorSubject.java:485)
     at io.reactivex.subjects.BehaviorSubject.onNext(BehaviorSubject.java:184)
     at com.trello.rxlifecycle2.components.support.RxFragment.onPause(RxFragment.java:88)
     at android.support.v4.app.Fragment.performPause(Fragment.java:2371)
     at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1357)
     at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1528)
     at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1595)
     at android.support.v4.app.FragmentManagerImpl.dispatchPause(FragmentManager.java:2920)
     at android.support.v4.app.FragmentController.dispatchPause(FragmentController.java:234)
     at android.support.v4.app.FragmentActivity.onPause(FragmentActivity.java:448)
     at com.trello.rxlifecycle2.components.support.RxAppCompatActivity.onPause(RxAppCompatActivity.java:81)
     at android.app.Activity.performPause(Activity.java:6819)
     at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1322)
     at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:3740)
     at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3717)
     at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3691)
     at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3665)
     at android.app.ActivityThread.-wrap16(ActivityThread.java)
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1489)
     at android.os.Handler.dispatchMessage(Handler.java:102)
     at android.os.Looper.loop(Looper.java:154)
     at android.app.ActivityThread.main(ActivityThread.java:6119)
     at java.lang.reflect.Method.invoke(Native Method)
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)

```

The issue can be seen in `SingleTakeUntil.TakeUntilMainObserver.otherError` which will throw an exception to the default UncaughtExceptionHandler if the Disposable(subscription) is disposed

How do I deal with this?

**Comment by dlew** [comment]
`Single` emits `CancellationException` when it disposes: https://github.com/trello/RxLifecycle#unsubscription

Due to the way RxLifecycle is setup, there isn't a great way around it. I suggest checking out [AutoDispose](https://github.com/uber/AutoDispose) if you want to automatically dispose of Singles without getting an exception.

**Comment by kassim** [comment]
Seems like it should come with a clearer warning in the documentation.

In any case, an easy workaround seems to be to convert your Single to an Observable

**Comment by masc3d** [comment]
I don't see how "no great way around it" would be a reason to close this though @dlew

**Comment by dlew** [comment]
Let me rephrase: there is *no way* around it since I'm not breaking anyone's existing code by changing behavior at this point.
